store/tikv/lock_resolver.go

```diff
@@ -578,10 +595,232 @@ func (lr *LockResolver) getTxnStatus(bo *Backoffer, txnID uint64, primary []byte
 			status.commitTS = cmdResp.CommitVersion
 			lr.saveResolved(txnID, status)
 		}
+
 		return status, nil
 	}
 }
 
+// asyncResolveData is data contributed by multiple goroutines when resolving locks using the async commit protocol. All
+// data should be protected by the mutex field.
+type asyncResolveData struct {
+	mutex       sync.Mutex
+	commitTs    uint64
+	locks       map[RegionVerID][]*Lock
+	missingLock bool
+}
+
+// addLocks adds locks to data, keeping other fields up to date. startTS and commitTS are for the transaction being
+// resolved.
+func (data *asyncResolveData) addLocks(locks []*kvrpcpb.LockInfo, curRegionID RegionVerID, expected int, startTS uint64, commitTS uint64) error {
+	data.mutex.Lock()
+	defer data.mutex.Unlock()
+
+	// Check locks to see if any have been committed or rolled back.
+	if len(locks) < expected {
+		// A lock is missing - the transaction must either be rolled back or committed.
+		if !data.missingLock {
+			data.commitTs = commitTS
+		}
+		data.missingLock = true
+
+		if data.commitTs != commitTS {
+			return errors.Errorf("commit TS mismatch in async commit recovery: %v and %v", data.commitTs, commitTS)
+		}
+	}
+
+	// Save all locks to be resolved.
```

Comment:
If `missingLock` is true, need we still do the following?

Reply:
I was unsure about this. I guess we don't because TiKV should have resolved all the locks in this region.

Reply:
Ah, makes sense. I am more convinced of returning all locks in check secondary keys. I will change the implementation in TiKV.